File: gauntlet/entries/tensor-map-4k/zig/main.zig
const std = @import("std");

const SIDE: usize = 4096;
const ELEMENTS: usize = SIDE * SIDE;

pub fn main(init: std.process.Init) !void {
    const allocator = std.heap.page_allocator;
    const left = try allocator.alloc(f64, ELEMENTS);
    defer allocator.free(left);
    const right = try allocator.alloc(f64, ELEMENTS);
    defer allocator.free(right);
    const mapped = try allocator.alloc(f64, ELEMENTS);
    defer allocator.free(mapped);

    for (left, right) |*a, *b| {
        a.* = 1.0;
        b.* = 2.0;
    }
    for (mapped, left, right) |*out, a, b| {
        out.* = a + b;
    }

    const corner = mapped[ELEMENTS - 1];
    var checksum: f64 = 0.0;
    for (mapped) |value| checksum += value;
    if (corner != 3.0 or checksum != 50331648.0) return error.InvalidResult;
    const checksum_int: u64 = @intFromFloat(checksum);

    var out_buffer: [128]u8 = undefined;
    var out = std.Io.File.Writer.init(.stdout(), init.io, &out_buffer);
    try out.interface.print(
        "shape:[4096, 4096] numel:{} corner:{d:.1} checksum:{}\n",
        .{ ELEMENTS, corner, checksum_int },
    );
    try out.interface.flush();
}

File: gauntlet/entries/tensor-map-4k/rust/main.rs
const SIDE: usize = 4096;
const ELEMENTS: usize = SIDE * SIDE;

fn main() {
    let left = vec![1.0_f64; ELEMENTS];
    let right = vec![2.0_f64; ELEMENTS];
    let mut mapped = Vec::with_capacity(ELEMENTS);
    for (a, b) in left.iter().zip(&right) {
        mapped.push(*a + *b);
    }

    let corner = mapped[ELEMENTS - 1];
    let checksum = mapped.iter().copied().sum::<f64>() as u64;
    assert_eq!(corner, 3.0);
    assert_eq!(checksum, 50_331_648);
    println!(
        "shape:[4096, 4096] numel:{ELEMENTS} corner:{corner:.1} checksum:{checksum}"
    );
}
